state/filesystem.go

```diff
@@ -588,13 +599,20 @@ func removeFilesystemAttachmentOps(m names.MachineTag, f *filesystem) []txn.Op {
 // DestroyFilesystem ensures that the filesystem and any attachments to it will
 // be destroyed and removed from state at some point in the future.
 func (st *State) DestroyFilesystem(tag names.FilesystemTag) (err error) {
+	defer errors.DeferredAnnotatef(&err, "destroying filesystem %s", tag.Id())
 	buildTxn := func(attempt int) ([]txn.Op, error) {
 		filesystem, err := st.filesystemByTag(tag)
 		if errors.IsNotFound(err) {
 			return nil, jujutxn.ErrNoOperations
 		} else if err != nil {
 			return nil, errors.Trace(err)
 		}
+		if filesystem.doc.StorageId != "" {
+			return nil, errors.Errorf(
+				"filesystem is assigned to %s",
```

Comment:
A filesystem is not necessarily attached to a machine, and is never directly associated with a unit. The associated storage could potentially be attached to multiple units. I'm going to leave this for now, unless you have strong feelings about it.